packages/websocket-plugin/src/websocket-handler.ts

```diff
@@ -14,45 +17,145 @@ import {
+  private setupActionsListeners(): void {
+    this.actions$.pipe(ofActionDispatched(ConnectWebSocket)).subscribe(({ payload }) => {
+      this.connect(payload);
     });
 
-    actions$.pipe(ofActionDispatched(DisconnectWebSocket)).subscribe(() => {
-      socket.disconnect();
+    this.actions$.pipe(ofActionDispatched(DisconnectWebSocket)).subscribe(() => {
+      this.disconnect();
     });
 
-    actions$.pipe(ofActionDispatched(SendWebSocketMessage)).subscribe(({ payload }) => {
-      socket.send(payload);
+    this.actions$.pipe(ofActionDispatched(SendWebSocketMessage)).subscribe(({ payload }) => {
+      this.send(payload);
     });
+  }
+
+  private connect(options?: NgxsWebsocketPluginOptions): void {
+    // To ensure we don't have any memory leaks
+    // e.g. if the user occasionally dispatched `ConnectWebSocket` twice
+    // then the previous subscription will still live in the memory
+    // to prevent such behavior - we close the previous connection if it exists
+    if (this.socket) {
+      this.socket.complete();
+    }
+
+    // Users can pass the options in the connect method so
+    // if options aren't available at DI bootstrap they have access
+    // to pass them here
+    if (options) {
+      this.mergeConfigWithOptions(options);
+    }
 
-    socket.subscribe(
-      message => {
-        const type = getValue(message, typeKey);
+    this.socket = new WebSocketSubject(this.config);
+
+    this.socket.subscribe({
+      next: (message: any) => {
+        const type = getValue(message, this.typeKey);
         if (!type) {
-          throw new TypeKeyPropertyMissingError(typeKey);
+          throw new TypeKeyPropertyMissingError(this.typeKey);
         }
-        store.dispatch({ ...message, type });
+        this.store.dispatch({ ...message, type });
       },
-      error => {
+      error: (error: any) => {
         if (error instanceof CloseEvent) {
-          store.dispatch(new WebSocketDisconnected());
+          this.dispatchWebSocketDisconnected();
         } else {
-          store.dispatch(new WebsocketMessageError(error));
+          this.store.dispatch(new WebsocketMessageError(error));
         }
       }
-    );
-
-    socket.completeWebSocket$.subscribe(() => {
-      store.dispatch(new WebSocketDisconnected());
     });
   }
+
+  private disconnect(): void {
+    if (this.socket) {
+      // `socket.complete()` closes the connection
+      // also it doesn't invoke the `onComplete` callback that we passed
+      // into `socket.subscribe(...)`
+      this.socket.complete();
+      this.socket = null;
+      this.dispatchWebSocketDisconnected();
+    }
+  }
+
+  private send(data: any): void {
+    if (!this.socket) {
+      throw new Error('You must connect to the socket before sending any data');
+    }
+
+    this.socket.next(data);
+  }
+
+  /**
+   * Don't enlarge the `connect` method
+   */
+  private mergeConfigWithOptions(options: NgxsWebsocketPluginOptions): void {
+    if (options.url) {
```

Comment:
This is subtly different because it is checking for values, as opposed to just overwriting them. @arturovt is this check deliberate? If so then keep the ifs, otherwise this is a good simplification of the code.

Reply:
but this check still does not make any sense, even if you pass any fields, they are still replaced with keys

Reply:
I guess the question is: if you explicitly pass an option as null or empty string would you want it to overwrite the config?

Reply:
This code is also saying that you can only change the `url`, `serialiser` or `deserialiser` in the config and not any of the other options. My point is that your suggested code of `this.config = { ...this.config, ...options };` is not logically equivalent. This code does something different, so this would be a behavior change.

If we do want to allow for the changing of any of the properties and potentially the setting of those properties to null, then your code would be the best option.

Reply:
Explicit `if` is a guard for the strictly predefined properties like `serializer`. We want to prevent the behavior when the user, due to carelessness, sets this property to `null`, the server side will also crash if the object is not serialized correctly.

Reply:
add comments please in code